File: backend/app/modules/identity.py
"""Identity module — auth, profile, org members, audit log (FR-USR).

Endpoints (mounted under /v1 by main.py):
- POST  /auth/register   create Organisation + admin User, return token immediately
- POST  /auth/login      audited ('auth.login'); failures are generic 401s
- GET   /me / PATCH /me  own profile
- GET   /members         (view)
- POST  /members/invite  (manage_members)
- PATCH /members/{id}    (manage_members) — cannot demote the last admin
- DELETE /members/{id}   (manage_members) — cannot delete yourself
- GET   /audit           (view_audit_log) — newest first, cursor pagination
- GET   /audit/export.csv, GET/PUT /audit/retention, POST /audit/purge  (FR-082)
"""
import json
import re
from datetime import datetime, timezone

from fastapi import APIRouter, Body, Depends, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field
from sqlalchemy import and_, func, or_
from sqlalchemy.orm import Session

from ..config import settings
from ..db import get_db
from ..deps import audit, get_current_user, require
from ..models import AuditEntry, Organisation, User
from ..security import ROLES, create_token, hash_password, verify_password

router = APIRouter()

_EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")
_LOCALES = ("en", "ar")


# --- helpers -----------------------------------------------------------------

def _iso(dt):
    return dt.isoformat() if dt else None


def _user_payload(user: User, org_name: str | None = None) -> dict:
    data = {
        "id": user.id,
        "name": user.name,
        "email": user.email,
        "role": user.role,
        "locale": user.locale,
        "organisation_id": user.organisation_id,
        "created_at": _iso(user.created_at),
    }
    if org_name is not None:
        data["org_name"] = org_name
    return data


def _org_name(db: Session, org_id: str) -> str:
    org = db.get(Organisation, org_id)
    return org.name if org else ""


def _validate_email(email: str) -> str:
    email = email.strip().lower()
    if not _EMAIL_RE.match(email):
        raise HTTPException(422, detail={"code": "invalid_email", "message": "Invalid email address"})
    return email


def _validate_role(role: str) -> str:
    if role not in ROLES:
        raise HTTPException(422, detail={
            "code": "invalid_role", "message": f"Role must be one of: {', '.join(ROLES)}"})
    return role


def _get_member(member_id: str, user: User, db: Session) -> User:
    """Org-isolated member lookup (FR-USR-04)."""
    member = db.get(User, member_id)
    if not member or member.organisation_id != user.organisation_id:
        raise HTTPException(404, detail={"code": "not_found", "message": "Member not found"})
    return member


def _admin_count(db: Session, org_id: str) -> int:
    return db.query(User).filter(
        User.organisation_id == org_id, User.role == "admin").count()


# --- request models ----------------------------------------------------------

class RegisterIn(BaseModel):
    org_name: str = Field(min_length=1, max_length=200)
    name: str = Field(min_length=1, max_length=200)
    email: str
    password: str = Field(min_length=8, max_length=200)
    locale: str = "en"


class LoginIn(BaseModel):
    email: str
    password: str


class MeUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    locale: str | None = None


class InviteIn(BaseModel):
    email: str
    name: str = Field(min_length=1, max_length=200)
    role: str = "qa_engineer"
    password: str = Field(min_length=8, max_length=200)


class RoleUpdate(BaseModel):
    role: str


# --- auth --------------------------------------------------------------------

@router.post("/auth/register", status_code=201)
def register(body: RegisterIn, db: Session = Depends(get_db)):
    email = _validate_email(body.email)
    if body.locale not in _LOCALES:
        raise HTTPException(422, detail={"code": "invalid_locale", "message": "Locale must be 'en' or 'ar'"})
    if db.query(User).filter(User.email == email).first():
        raise HTTPException(409, detail={"code": "email_taken", "message": "Email is already registered"})

    org = Organisation(name=body.org_name.strip())
    db.add(org)
    db.flush()
    user = User(organisation_id=org.id, email=email, name=body.name.strip(),
                password_hash=hash_password(body.password), role="admin", locale=body.locale)
    db.add(user)
    db.flush()
    audit(db, org.id, user.id, "auth.register", "user", user.id,
          {"email": email, "org_name": org.name})
    db.commit()
    return {"token": create_token(user.id, org.id, user.role),
            "user": _user_payload(user, org.name)}


@router.post("/auth/login")
def login(body: LoginIn, db: Session = Depends(get_db)):
    email = body.email.strip().lower()
    user = db.query(User).filter(User.email == email).first()
    if not user or not verify_password(body.password, user.password_hash):
        # Never reveal which field was wrong (NFR-SEC).
        if user:
            audit(db, user.organisation_id, user.id, "auth.login_failed", "user", user.id,
                  {"email": email})
            db.commit()
        raise HTTPException(401, detail={"code": "invalid_credentials",
                                         "message": "Invalid email or password"})
    audit(db, user.organisation_id, user.id, "auth.login", "user", user.id, {"email": email})
    db.commit()
    return {"token": create_token(user.id, user.organisation_id, user.role),
            "user": _user_payload(user, _org_name(db, user.organisation_id))}


# --- own profile -------------------------------------------------------------

@router.get("/me")
def get_me(user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return _user_payload(user, _org_name(db, user.organisation_id))


@router.patch("/me")
def update_me(body: MeUpdate, user: User = Depends(get_current_user),
              db: Session = Depends(get_db)):
    if body.name is not None:
        user.name = body.name.strip()
    if body.locale is not None:
        if body.locale not in _LOCALES:
            raise HTTPException(422, detail={"code": "invalid_locale",
                                             "message": "Locale must be 'en' or 'ar'"})
        user.locale = body.locale
    db.commit()
    return _user_payload(user, _org_name(db, user.organisation_id))


# --- members -----------------------------------------------------------------

@router.get("/members")
def list_members(user: User = Depends(require("view")), db: Session = Depends(get_db)):
    members = (db.query(User)
               .filter(User.organisation_id == user.organisation_id)
               .order_by(User.created_at.asc())
               .all())
    return [_user_payload(m) for m in members]


@router.post("/members/invite", status_code=201)
def invite_member(body: InviteIn, user: User = Depends(require("manage_members")),
                  db: Session = Depends(get_db)):
    email = _validate_email(body.email)
    _validate_role(body.role)
    if db.query(User).filter(User.email == email).first():
        raise HTTPException(409, detail={"code": "email_taken", "message": "Email is already registered"})
    member = User(organisation_id=user.organisation_id, email=email, name=body.name.strip(),
                  password_hash=hash_password(body.password), role=body.role)
    db.add(member)
    db.flush()
    audit(db, user.organisation_id, user.id, "member.invite", "user", member.id,
          {"email": email, "role": body.role})
    db.commit()
    return _user_payload(member)


@router.patch("/members/{member_id}")
def update_member(member_id: str, body: RoleUpdate,
                  user: User = Depends(require("manage_members")),
                  db: Session = Depends(get_db)):
    _validate_role(body.role)
    member = _get_member(member_id, user, db)
    if member.role == "admin" and body.role != "admin" and \
            _admin_count(db, user.organisation_id) <= 1:
        raise HTTPException(400, detail={"code": "last_admin",
                                         "message": "Cannot demote the last admin of the organisation"})
    old_role = member.role
    member.role = body.role
    audit(db, user.organisation_id, user.id, "member.role_change", "user", member.id,
          {"from": old_role, "to": body.role})
    db.commit()
    return _user_payload(member)


@router.delete("/members/{member_id}")
def delete_member(member_id: str, user: User = Depends(require("manage_members")),
                  db: Session = Depends(get_db)):
    member = _get_member(member_id, user, db)
    if member.id == user.id:
        raise HTTPException(400, detail={"code": "cannot_delete_self",
                                         "message": "You cannot delete your own account"})
    if member.role == "admin" and _admin_count(db, user.organisation_id) <= 1:
        raise HTTPException(400, detail={"code": "last_admin",
                                         "message": "Cannot delete the last admin of the organisation"})
    audit(db, user.organisation_id, user.id, "member.delete", "user", member.id,
          {"email": member.email})
    db.delete(member)
    db.commit()
    return {"deleted": True}


# --- audit log ---------------------------------------------------------------

@router.get("/audit")
def audit_log(limit: int = 50, cursor: str | None = None,
              user: User = Depends(require("view_audit_log")),
              db: Session = Depends(get_db)):
    limit = max(1, min(limit, 200))
    q = db.query(AuditEntry).filter(AuditEntry.organisation_id == user.organisation_id)
    if cursor:
        anchor = db.get(AuditEntry, cursor)
        if not anchor or anchor.organisation_id != user.organisation_id:
            raise HTTPException(400, detail={"code": "invalid_cursor", "message": "Unknown cursor"})
        q = q.filter(or_(AuditEntry.occurred_at < anchor.occurred_at,
                         and_(AuditEntry.occurred_at == anchor.occurred_at,
                              AuditEntry.id < anchor.id)))
    entries = (q.order_by(AuditEntry.occurred_at.desc(), AuditEntry.id.desc())
               .limit(limit + 1).all())
    has_more = len(entries) > limit
    items = entries[:limit]
    return {
        "items": [{
            "id": e.id,
            "actor_id": e.actor_id,
            "action": e.action,
            "object_type": e.object_type,
            "object_id": e.object_id,
            "detail": e.detail or {},
            "occurred_at": _iso(e.occurred_at),
            "retain_until": _iso(e.retain_until),
        } for e in items],
        "next_cursor": items[-1].id if has_more and items else None,
    }


# --- audit retention + export (FR-082 AC2/AC3/AC4) ----------------------------

@router.get("/audit/retention")
def get_retention(user: User = Depends(require("view_audit_log")),
                  db: Session = Depends(get_db)):
    org = db.get(Organisation, user.organisation_id)
    days = int((org.settings or {}).get("audit_retention_days") or settings.AUDIT_RETENTION_DAYS)
    purgeable = (db.query(func.count(AuditEntry.id))
                 .filter(AuditEntry.organisation_id == user.organisation_id,
                         AuditEntry.retain_until != None,  # noqa: E711
                         AuditEntry.retain_until <= datetime.now(timezone.utc))
                 .scalar() or 0)
    total = (db.query(func.count(AuditEntry.id))
             .filter(AuditEntry.organisation_id == user.organisation_id).scalar() or 0)
    return {"retention_days": days, "default_days": settings.AUDIT_RETENTION_DAYS,
            "entries": total, "past_retention": purgeable}


@router.put("/audit/retention")
def set_retention(body: dict = Body(...),
                  user: User = Depends(require("manage_members")),
                  db: Session = Depends(get_db)):
    days = body.get("retention_days")
    if not isinstance(days, int) or not (1 <= days <= 3650):
        raise HTTPException(422, detail={
            "code": "invalid_retention",
            "message": "retention_days must be an integer between 1 and 3650"})
    org = db.get(Organisation, user.organisation_id)
    org.settings = {**(org.settings or {}), "audit_retention_days": days}
    audit(db, user.organisation_id, user.id, "audit.retention_changed", "organisation",
          org.id, {"retention_days": days})
    db.commit()
    return {"retention_days": days}


@router.post("/audit/purge")
def purge_expired(user: User = Depends(require("manage_members")),
                  db: Session = Depends(get_db)):
    """The ONLY deletion path, and it cannot touch an entry before its retention date
    (FR-082 AC2). Entries are never editable through any route."""
    now = datetime.now(timezone.utc)
    removed = (db.query(AuditEntry)
               .filter(AuditEntry.organisation_id == user.organisation_id,
                       AuditEntry.retain_until != None,  # noqa: E711
                       AuditEntry.retain_until <= now)
               .delete(synchronize_session=False))
    audit(db, user.organisation_id, user.id, "audit.purged", "organisation",
          user.organisation_id, {"removed": removed})
    db.commit()
    return {"removed": removed}


@router.get("/audit/export.csv")
def export_audit(user: User = Depends(require("view_audit_log")),
                 db: Session = Depends(get_db)):
    """FR-082 AC4 — the whole log, in the format an auditor asks for."""
    import csv
    import io

    buf = io.StringIO()
    writer = csv.writer(buf)
    writer.writerow(["id", "occurred_at", "actor_id", "action", "object_type",
                     "object_id", "retain_until", "detail"])
    rows = (db.query(AuditEntry)
            .filter(AuditEntry.organisation_id == user.organisation_id)
            .order_by(AuditEntry.occurred_at.asc(), AuditEntry.id.asc()).all())
    for e in rows:
        writer.writerow([e.id, _iso(e.occurred_at), e.actor_id or "", e.action,
                         e.object_type, e.object_id, _iso(e.retain_until) or "",
                         json.dumps(e.detail or {}, ensure_ascii=False)])
    buf.seek(0)
    return StreamingResponse(
        iter([buf.getvalue()]), media_type="text/csv; charset=utf-8",
        headers={"Content-Disposition": 'attachment; filename="traceo-audit.csv"'})

File: backend-go/internal/modules/identity/identity.go
// Package identity — auth, profile, org members, audit log (FR-USR).
// 1:1 port of backend/app/modules/identity.py.
//
// Endpoints (mounted under /v1):
//   - POST   /auth/register   create Organisation + admin User, return token immediately
//   - POST   /auth/login      audited ('auth.login'); failures are generic 401s
//   - POST   /auth/dev-session  development-only credential-free session; 404 unless
//     TRACEO_DEV_AUTOLOGIN=1 (audited 'auth.dev_session')
//   - GET    /me / PATCH /me  own profile
//   - GET    /members         (view)
//   - POST   /members/invite  (manage_members)
//   - PATCH  /members/{id}    (manage_members) — cannot demote the last admin
//   - DELETE /members/{id}    (manage_members) — cannot delete yourself
//   - GET    /audit           (view_audit_log) — newest first, cursor pagination
package identity

import (
	"net/http"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/gin-gonic/gin"

	"traceo/internal/config"
	"traceo/internal/db"
	"traceo/internal/httpx"
	"traceo/internal/models"
	"traceo/internal/security"
)

var (
	emailRe = regexp.MustCompile(`^[^@\s]+@[^@\s]+\.[^@\s]+$`)
	roles   = []string{"admin", "qa_lead", "qa_engineer", "viewer"}
	locales = []string{"en", "ar"}
)

// --- helpers -----------------------------------------------------------------

func iso(t time.Time) string { return t.UTC().Format(time.RFC3339) }

func userPayload(u *models.User) gin.H {
	return gin.H{
		"id":              u.ID,
		"name":            u.Name,
		"email":           u.Email,
		"role":            u.Role,
		"locale":          u.Locale,
		"organisation_id": u.OrganisationID,
		"created_at":      iso(u.CreatedAt),
	}
}

func userPayloadWithOrg(u *models.User, orgName string) gin.H {
	p := userPayload(u)
	p["org_name"] = orgName
	return p
}

func orgName(orgID string) string {
	var org models.Organisation
	if err := db.DB.First(&org, "id = ?", orgID).Error; err != nil {
		return ""
	}
	return org.Name
}

// validateEmail normalises and checks the address; writes 422 + returns ok=false on failure.
func validateEmail(c *gin.Context, email string) (string, bool) {
	email = strings.ToLower(strings.TrimSpace(email))
	if !emailRe.MatchString(email) {
		httpx.Err(c, http.StatusUnprocessableEntity, "invalid_email", "Invalid email address")
		return "", false
	}
	return email, true
}

func validateRole(c *gin.Context, role string) bool {
	for _, r := range roles {
		if r == role {
			return true
		}
	}
	httpx.Err(c, http.StatusUnprocessableEntity, "invalid_role",
		"Role must be one of: "+strings.Join(roles, ", "))
	return false
}

func validLocale(l string) bool { return l == "en" || l == "ar" }

// getMember does the org-isolated member lookup (FR-USR-04); writes 404 on failure.
func getMember(c *gin.Context, memberID string, u *models.User) (*models.User, bool) {
	var member models.User
	if err := db.DB.First(&member, "id = ?", memberID).Error; err != nil ||
		member.OrganisationID != u.OrganisationID {
		httpx.Err(c, http.StatusNotFound, "not_found", "Member not found")
		return nil, false
	}
	return &member, true
}

func adminCount(orgID string) int64 {
	var n int64
	db.DB.Model(&models.User{}).Where("organisation_id = ? AND role = ?", orgID, "admin").Count(&n)
	return n
}

func emailTaken(email string) bool {
	var n int64
	db.DB.Model(&models.User{}).Where("email = ?", email).Count(&n)
	return n > 0
}

// --- routes ------------------------------------------------------------------

func Register(r *gin.RouterGroup) {
	r.POST("/auth/register", register)
	r.POST("/auth/login", login)
	r.POST("/auth/dev-session", devSession)

	r.GET("/me", httpx.Auth(), getMe)
	r.PATCH("/me", httpx.Auth(), updateMe)

	r.GET("/members", httpx.Auth(), httpx.Require("view"), listMembers)
	r.POST("/members/invite", httpx.Auth(), httpx.Require("manage_members"), inviteMember)
	r.PATCH("/members/:member_id", httpx.Auth(), httpx.Require("manage_members"), updateMember)
	r.DELETE("/members/:member_id", httpx.Auth(), httpx.Require("manage_members"), deleteMember)

	r.GET("/audit", httpx.Auth(), httpx.Require("view_audit_log"), auditLog)
}

// --- auth --------------------------------------------------------------------

func register(c *gin.Context) {
	var body struct {
		OrgName  string  `json:"org_name"`
		Name     string  `json:"name"`
		Email    string  `json:"email"`
		Password string  `json:"password"`
		Locale   *string `json:"locale"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		httpx.Err(c, http.StatusUnprocessableEntity, "validation_error", "Invalid request body")
		return
	}
	if len(body.OrgName) < 1 || len(body.OrgName) > 200 ||
		len(body.Name) < 1 || len(body.Name) > 200 ||
		len(body.Password) < 8 || len(body.Password) > 200 {
		httpx.Err(c, http.StatusUnprocessableEntity, "validation_error", "Invalid field length")
		return
	}
	email, ok := validateEmail(c, body.Email)
	if !ok {
		return
	}
	locale := "en"
	if body.Locale != nil {
		locale = *body.Locale
	}
	if !validLocale(locale) {
		httpx.Err(c, http.StatusUnprocessableEntity, "invalid_locale", "Locale must be 'en' or 'ar'")
		return
	}
	if emailTaken(email) {
		httpx.Err(c, http.StatusConflict, "email_taken", "Email is already registered")
		return
	}

	hash, err := security.HashPassword(body.Password)
	if err != nil {
		httpx.Err(c, http.StatusInternalServerError, "internal_error", "Could not hash password")
		return
	}
	org := models.Organisation{Name: strings.TrimSpace(body.OrgName)}
	if err := db.DB.Create(&org).Error; err != nil {
		httpx.Err(c, http.StatusInternalServerError, "internal_error", "Could not create organisation")
		return
	}
	user := models.User{
		OrganisationID: org.ID, Email: email, Name: strings.TrimSpace(body.Name),
		PasswordHash: hash, Role: "admin", Locale: locale,
	}
	if err := db.DB.Create(&user).Error; err != nil {
		httpx.Err(c, http.StatusInternalServerError, "internal_error", "Could not create user")
		return
	}
	httpx.Audit(org.ID, &user.ID, "auth.register", "user", user.ID,
		models.JSONMap{"email": email, "org_name": org.Name})
	token, _ := security.CreateToken(user.ID, org.ID, user.Role)
	c.JSON(http.StatusCreated, gin.H{"token": token, "user": userPayloadWithOrg(&user, org.Name)})
}

func login(c *gin.Context) {
	var body struct {
		Email    string `json:"email"`
		Password string `json:"password"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		httpx.Err(c, http.StatusUnprocessableEntity, "validation_error", "Invalid request body")
		return
	}
	email := strings.ToLower(strings.TrimSpace(body.Email))
	var user models.User
	found := db.DB.First(&user, "email = ?", email).Error == nil
	if !found || !security.VerifyPassword(body.Password, user.PasswordHash) {
		// Never reveal which field was wrong (NFR-SEC).
		if found {
			httpx.Audit(user.OrganisationID, &user.ID, "auth.login_failed", "user", user.ID,
				models.JSONMap{"email": email})
		}
		httpx.Err(c, http.StatusUnauthorized, "invalid_credentials", "Invalid email or password")
		return
	}
	httpx.Audit(user.OrganisationID, &user.ID, "auth.login", "user", user.ID,
		models.JSONMap{"email": email})
	token, _ := security.CreateToken(user.ID, user.OrganisationID, user.Role)
	c.JSON(http.StatusOK, gin.H{"token": token,
		"user": userPayloadWithOrg(&user, orgName(user.OrganisationID))})
}

// devSession hands out a session without credentials — development only.
//
// Enabled by TRACEO_DEV_AUTOLOGIN=1; the production guard in config refuses to
// boot with it set, so this can never be reachable on a production node. While
// the flag is off the route is indistinguishable from a missing one (404), so a
// misconfigured deployment leaks nothing about the feature's existence.
func devSession(c *gin.Context) {
	if !config.C.DevAutologin {
		httpx.Err(c, http.StatusNotFound, "not_found", "Not found")
		return
	}
	email := strings.ToLower(strings.TrimSpace(config.C.DevAutologinEmail))
	var user models.User
	if err := db.DB.First(&user, "email = ?", email).Error; err != nil {
		httpx.Err(c, http.StatusServiceUnavailable, "dev_session_unavailable",
			"TRACEO_DEV_AUTOLOGIN is on but no user matches "+email)
		return
	}
	httpx.Audit(user.OrganisationID, &user.ID, "auth.dev_session", "user", user.ID,
		models.JSONMap{"email": email})
	token, _ := security.CreateToken(user.ID, user.OrganisationID, user.Role)
	c.JSON(http.StatusOK, gin.H{"token": token,
		"user": userPayloadWithOrg(&user, orgName(user.OrganisationID))})
}

// --- own profile -------------------------------------------------------------

func getMe(c *gin.Context) {
	u := httpx.User(c)
	c.JSON(http.StatusOK, userPayloadWithOrg(u, orgName(u.OrganisationID)))
}

func updateMe(c *gin.Context) {
	u := httpx.User(c)
	var body struct {
		Name   *string `json:"name"`
		Locale *string `json:"locale"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		httpx.Err(c, http.StatusUnprocessableEntity, "validation_error", "Invalid request body")
		return
	}
	if body.Name != nil {
		if len(*body.Name) < 1 || len(*body.Name) > 200 {
			httpx.Err(c, http.StatusUnprocessableEntity, "validation_error", "Invalid field length")
			return
		}
		u.Name = strings.TrimSpace(*body.Name)
	}
	if body.Locale != nil {
		if !validLocale(*body.Locale) {
			httpx.Err(c, http.StatusUnprocessableEntity, "invalid_locale", "Locale must be 'en' or 'ar'")
			return
		}
		u.Locale = *body.Locale
	}
	db.DB.Save(u)
	c.JSON(http.StatusOK, userPayloadWithOrg(u, orgName(u.OrganisationID)))
}

// --- members -----------------------------------------------------------------

func listMembers(c *gin.Context) {
	u := httpx.User(c)
	var members []models.User
	db.DB.Where("organisation_id = ?", u.OrganisationID).
		Order("created_at asc").Find(&members)
	out := make([]gin.H, 0, len(members))
	for i := range members {
		out = append(out, userPayload(&members[i]))
	}
	c.JSON(http.StatusOK, out)
}

func inviteMember(c *gin.Context) {
	u := httpx.User(c)
	var body struct {
		Email    string  `json:"email"`
		Name     string  `json:"name"`
		Role     *string `json:"role"`
		Password string  `json:"password"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		httpx.Err(c, http.StatusUnprocessableEntity, "validation_error", "Invalid request body")
		return
	}
	if len(body.Name) < 1 || len(body.Name) > 200 ||
		len(body.Password) < 8 || len(body.Password) > 200 {
		httpx.Err(c, http.StatusUnprocessableEntity, "validation_error", "Invalid field length")
		return
	}
	email, ok := validateEmail(c, body.Email)
	if !ok {
		return
	}
	role := "qa_engineer"
	if body.Role != nil {
		role = *body.Role
	}
	if !validateRole(c, role) {
		return
	}
	if emailTaken(email) {
		httpx.Err(c, http.StatusConflict, "email_taken", "Email is already registered")
		return
	}
	hash, err := security.HashPassword(body.Password)
	if err != nil {
		httpx.Err(c, http.StatusInternalServerError, "internal_error", "Could not hash password")
		return
	}
	member := models.User{
		OrganisationID: u.OrganisationID, Email: email, Name: strings.TrimSpace(body.Name),
		PasswordHash: hash, Role: role,
	}
	if err := db.DB.Create(&member).Error; err != nil {
		httpx.Err(c, http.StatusInternalServerError, "internal_error", "Could not create user")
		return
	}
	httpx.Audit(u.OrganisationID, &u.ID, "member.invite", "user", member.ID,
		models.JSONMap{"email": email, "role": role})
	c.JSON(http.StatusCreated, userPayload(&member))
}

func updateMember(c *gin.Context) {
	u := httpx.User(c)
	var body struct {
		Role string `json:"role"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		httpx.Err(c, http.StatusUnprocessableEntity, "validation_error", "Invalid request body")
		return
	}
	if !validateRole(c, body.Role) {
		return
	}
	member, ok := getMember(c, c.Param("member_id"), u)
	if !ok {
		return
	}
	if member.Role == "admin" && body.Role != "admin" && adminCount(u.OrganisationID) <= 1 {
		httpx.Err(c, http.StatusBadRequest, "last_admin",
			"Cannot demote the last admin of the organisation")
		return
	}
	oldRole := member.Role
	member.Role = body.Role
	db.DB.Save(member)
	httpx.Audit(u.OrganisationID, &u.ID, "member.role_change", "user", member.ID,
		models.JSONMap{"from": oldRole, "to": body.Role})
	c.JSON(http.StatusOK, userPayload(member))
}

func deleteMember(c *gin.Context) {
	u := httpx.User(c)
	member, ok := getMember(c, c.Param("member_id"), u)
	if !ok {
		return
	}
	if member.ID == u.ID {
		httpx.Err(c, http.StatusBadRequest, "cannot_delete_self",
			"You cannot delete your own account")
		return
	}
	if member.Role == "admin" && adminCount(u.OrganisationID) <= 1 {
		httpx.Err(c, http.StatusBadRequest, "last_admin",
			"Cannot delete the last admin of the organisation")
		return
	}
	httpx.Audit(u.OrganisationID, &u.ID, "member.delete", "user", member.ID,
		models.JSONMap{"email": member.Email})
	db.DB.Delete(&models.User{}, "id = ?", member.ID)
	c.JSON(http.StatusOK, gin.H{"deleted": true})
}

// --- audit log ---------------------------------------------------------------

func auditLog(c *gin.Context) {
	u := httpx.User(c)
	limit, err := strconv.Atoi(c.DefaultQuery("limit", "50"))
	if err != nil {
		httpx.Err(c, http.StatusUnprocessableEntity, "validation_error", "limit must be an integer")
		return
	}
	if limit < 1 {
		limit = 1
	}
	if limit > 200 {
		limit = 200
	}

	q := db.DB.Model(&models.AuditEntry{}).Where("organisation_id = ?", u.OrganisationID)
	if cursor := c.Query("cursor"); cursor != "" {
		var anchor models.AuditEntry
		if err := db.DB.First(&anchor, "id = ?", cursor).Error; err != nil ||
			anchor.OrganisationID != u.OrganisationID {
			httpx.Err(c, http.StatusBadRequest, "invalid_cursor", "Unknown cursor")
			return
		}
		q = q.Where("occurred_at < ? OR (occurred_at = ? AND id < ?)",
			anchor.OccurredAt, anchor.OccurredAt, anchor.ID)
	}
	var entries []models.AuditEntry
	q.Order("occurred_at desc, id desc").Limit(limit + 1).Find(&entries)

	hasMore := len(entries) > limit
	if hasMore {
		entries = entries[:limit]
	}
	items := make([]gin.H, 0, len(entries))
	for _, e := range entries {
		detail := e.Detail
		if detail == nil {
			detail = models.JSONMap{}
		}
		items = append(items, gin.H{
			"id":          e.ID,
			"actor_id":    e.ActorID,
			"action":      e.Action,
			"object_type": e.ObjectType,
			"object_id":   e.ObjectID,
			"detail":      detail,
			"occurred_at": iso(e.OccurredAt),
		})
	}
	var nextCursor any
	if hasMore && len(entries) > 0 {
		nextCursor = entries[len(entries)-1].ID
	}
	c.JSON(http.StatusOK, gin.H{"items": items, "next_cursor": nextCursor})
}
